### ISSUE TITLE
When C# script extends Resource class, the resource doesn't show up in the "Create New Resource" menu

### ISSUE THREAD

**Issue by sudoman281** [godotengine/godot#27470]
**Godot version:**
3.1

**OS/device including version:**
Arch Linux - newest

**Issue description:**
When C# script extends Resource class, the resource doesn't show up in the "Create New Resource" menu

**Steps to reproduce:**
Create a C# script, make it extend Resource class and then try to create a new resource of that type.

**Comment by PanzerRavana** [comment]
Please add this, people looking to swith from unity will have a much easier time adapting to godot

**Comment by akien-mga** [comment]
Extending Resource doesn't automatically register it as a resource usable in the editor.

`class_name` can be used in GDScript to register node types, but I don't know if it works for resources. There's no equivalent for C# yet that I know of.

**Comment by CJNA** [comment]
I have tackled this issue and this problem is harder than I think. Currently, there exists circular dependency between resourcePreview and Resource class, and thus at least in editor, it is unlikely to implement a way to bypass this circular dependency and add a generator on constructor of Resource.

Unless there is other way of calling resourcePreview when extending Resource class, it is required to restructure resource class implementation and hierachy to resolve this issue.

**Comment by scott-lin** [comment]
Is there a workaround for C# defined resources in the meantime?

**Comment by mysticfall** [comment]
I believe you can create a generic resource within Godot then attach your C# source which extends `Resouce` to the created asset.

**Comment by Spartan322** [comment]
Doesn't appear assigning a C# source to the Resource actually allows it to be edited.

**Comment by Reneator** [comment]
Maybe this issue is relevant:
https://github.com/godotengine/godot-proposals/issues/18#issuecomment-561864596

With a manually compiled godot project with the pull request by willnationsdev i am currently able to work with custom-resources in gdscript, but I am thinking about switching to C# (as its a language with features i might need for my project)

He stated that he has a C# implementation of it in the works.

**Comment by willnationsdev** [comment]
I have a WIP implementation of this ready-to-go (the latest commit on [this branch of my fork](https://github.com/willnationsdev/godot/tree/cs-script-classes)), but I've hit a roadblock in that the `ScriptClass` class attribute I've added never gets detected by the CSharpScript class. I've asked for @neikeq's help on the matter previously, but I never followed up when he didn't reply because I saw that he was busy trying to work on Android/Web exports for the 3.2 release and didn't want to bother him.

If anyone can take the work I've done in that commit and figure out how to make C# actually parse and detect the attribute, then I can continue developing it and hopefully finish it for godotengine/godot-proposals#22.

**Comment by vitorbalbio** [comment]
This is a huge limitation of C# and force us to change code structs to workaround it. How's the development on it issue?

**Comment by vitorbalbio** [comment]
For anyone how is stuck on this one workaround is to declare the class in GDScript extending Resources and call it in C#
example:

// data.gd
extends Resource
class_name Model_test 
var Speed : int = 0;
var Size : int = 0;

// UseData.cs
[Export] public List<Resource> dataList; // set data in inspector

**Comment by Calinou** [comment]
> This is a huge limitation of C# and force us to change code structs to workaround it. How's the development on it issue?

There's no ETA for fixing this issue, as nobody has made a fully functional fix yet.

**Comment by vitorbalbio** [comment]
> > This is a huge limitation of C# and force us to change code structs to workaround it. How's the development on it issue?
> 
> There's no ETA for fixing this issue, as nobody has made a fully functional fix yet.

That's sad but thanks for reply.

**Comment by willnationsdev** [comment]
@vitorbalbio Just an FYI, I've been gradually putting together a branch on my fork to implement various parts of godotengine/godot-proposals#22 (my `P22/script-classes` branch). The C# part hasn't been done yet, but I've updated my ScriptServer singleton inclusion and my VisualScript script class support. Just need to take the time to get a Mono build going again and see if I can't get things working in 4.0 (I never managed to make it work in 3.2 either since the script class attribute would never get detected. But we'll see what happens).

**Comment by vitorbalbio** [comment]
> @vitorbalbio Just an FYI, I've been gradually putting together a branch on my fork to implement various parts of [godotengine/godot-proposals#22](https://github.com/godotengine/godot-proposals/issues/22) (my `P22/script-classes` branch). The C# part hasn't been done yet, but I've updated my ScriptServer singleton inclusion and my VisualScript script class support. Just need to take the time to get a Mono build going again and see if I can't get things working in 4.0 (I never managed to make it work in 3.2 either since the script class attribute would never get detected. But we'll see what happens).

Hi willnationsdev! Thanks for your hard work, i'm a game developer working on indie studios in Brazil for about a decade (with Unity) and now we're giving a chance to Godot. Most of what we do can be done in godot (C# support is a must) but some paper cuts like this are stopping us to make a full jump on board. I will keep following Godot close and see how things evolve with time. Thank you.

**Comment by vitorbalbio** [comment]
> @vitorbalbio Just curious why just creating a generic resource and manually adding the resource script to the tres file not working for you?

It kinda works and is what i'm using mostly of the time. But once your data struct is too heavy you will need tons of tres files in your project. In my case: I have files for "units" (spaceships) each unity have "skills", each skill have "pattern". For each unity i have something like 15 tres. My game will have something like 20 units so hundreds of tres files to manage while it would be just 20 files if i was using GDScript.

**Comment by willnationsdev** [comment]
@vitorbalbio You should be able to create Resource properties ON resources which you can attach scripts to. Then, when you load the resource, it also loads the other resources embedded within it.

So, e.g. you could have a SpaceshipData resource that has exported Resource properties. You create a resource in the Inspector, add the SpaceshipData script to it. Suddenly, those exported resource properties (for, e.g. skills, pattern, etc.) are displayed. Then you create new resources in each of those and drag the respective resource scripts onto each of those properties (SkillData, PatternData, etc.). The sub-inspectors for those properties now have the scripted properties displayed. Once you have everything configured, you save the resource as a single file. And now, if you do `GD.Load("spaceship1.tres")`, you end up with a full SpaceshipData that contains all of the embedded resources.

**Comment by vitorbalbio** [comment]
**"You should be able to create Resource properties ON resources"**
I can

**"which you can attach scripts to. Then, when you load the resource"**
I can't. This option is not exposed

**"You create a resource in the Inspector"**
I can

**"add the SpaceshipData script to it."**
I can't. This option is not exposed.

See:
![Godot_v3 2 2-stable_mono_win64_rEB4432HZM](https://user-images.githubusercontent.com/6172013/86158799-b1b31b80-badf-11ea-93fa-2fae1bfc3ec4.png)
data is: "[Export] public Resource data;"

I can indeed set it as a new Resource in inspector:
![Godot_v3 2 2-stable_mono_win64_es7PNtRmDb](https://user-images.githubusercontent.com/6172013/86159139-3dc54300-bae0-11ea-83bd-35eb72c95591.png)

But after that i can't set any script or class to it:
![Godot_v3 2 2-stable_mono_win64_CSDL3UKxq9](https://user-images.githubusercontent.com/6172013/86159132-3bfb7f80-bae0-11ea-8475-5e9ff14e79ec.png)

**Comment by willnationsdev** [comment]
@vitorbalbio Oh, wat. That's weird. If you make a Resource individually in the Inspector, the `script` property shows up under a `Reference` category section. But that entire category section seems to be omitted in that dropdown. That feels like it should be considered a bug. All one would have to do is make sure that the script shows up in there and then you can assign a value there again.

**Comment by vitorbalbio** [comment]
**"If you make a Resource individually in the Inspector, the script property shows up under a Reference category section."**
This is exactly how i create the data files. I add a new "Resource" in the project and set the script in inspector. But as you can see i can't do it without create a new file first. By the end i will have a lot of .tres data files.

I don't think its a bug either since you can't select the script for any "sub inspector" in godot. But it would be great if only it was exposed in inspector!

**Comment by willnationsdev** [comment]
> I don't think its a bug either since you can't select the script for any "sub inspector" in godot.

Doesn't mean we shouldn't interpret this as a bug, hehe. At least, **I** would see it as a bug. It's silly that you can change the type of a resource at the top level, but not in a nested level.

**Comment by vitorbalbio** [comment]
I would love to see this implemented anyway ;)

**Comment by SimplyMiha** [comment]
What is the status of C# custom resources?

**Comment by Calinou** [comment]
@SimplyMiha As far as I know, nobody has started implementing this yet.

**Comment by willnationsdev** [comment]
@SimplyMiha the missing pieces are my two existing PRs that are both WIP (one for custom resources export, one for adding script class support to other languages). However, despite that, I've been having trouble even getting a functioning mono build of the master branch on Windows. Every time I build it and run, the engine closes almost immediately. Can't even test whether my class attribute is working or not.

Anyway, once we have script class support on C# classes and the ability to export custom resources *in general*, then we will may need an additional PR to ensure that exporting a script class named C# resource works in the C# language.

**Comment by vitorbalbio** [comment]
> @SimplyMiha the missing pieces are my two existing PRs that are both WIP (one for custom resources export, one for adding script class support to other languages). However, despite that, I've been having trouble even getting a functioning mono build of the master branch on Windows. Every time I build it and run, the engine closes almost immediately. Can't even test whether my class attribute is working or not.
> 
> Anyway, once we have script class support on C# classes and the ability to export custom resources _in general_, then we will may need an additional PR to ensure that exporting a script class named C# resource works in the C# language.

This looks a lot of work. There's one contributor working exclusively in C# support right? Maybe someone can poke him to help about it.

**Comment by Neotoon** [comment]
https://github.com/godotengine/godot/issues/15210
recreated the steps they did here and was able to create a c# script that extends resource class.

P.S if this comment violates any rules please let me know and kindly remove this.

**Comment by rob-mur** [comment]
Stumbled across this and found a "somewhat" ok workaround. The idea is to first define your custom resources using gdscript, and then use a generic helper method to read these in to their equivalent c# objects using reflection.

I've attached a example project that lets you read in a nested custom resource. It's currently just a prototype so you'd have to write the "other way" i.e. saving to a resource, but should be enough to go on for a current workaround. Ideally of course this wouldn't be needed!
[reflection test.zip](https://github.com/godotengine/godot/files/5682987/reflection.test.zip)

**Comment by Atlinx** [comment]
I spent some time digging around online and found a very clean Plugin solution written by [wgmor](https://github.com/wmigor) which solves this issue. I ended up modifying it a little bit to support registering custom C# resources anywhere under a specified list of directories to check, and it works perfectly for my needs.

If anyone is interested in using it, here the repo:
[Godot-Mono-CustomResourceRegister](https://github.com/Atlinx/Godot-Mono-CustomResourceRegister)

**Comment by cgbeutler** [comment]
Here is another little workaround that's fairly simple.
Hit the little refresh next to a resource to get a new one. It's not pretty, but it may hold you over.
```csharp
private CustomResource __res = null;
[Export]
public CustomResource res {
    get {
        if ( __res == null )  res = null; // lazy init (in-case godot loads from disk instead.)
        return __res;
    }
    set {
        if ( value == null ) {
            value = GD.Load<CSharpScript>("res://CustomResource.cs").New() as CustomResource;
        }
        __res = value;
    }
}
```
(Note: You can't make custom resources directly in C#, as their script seems to lack the resource_path needed to save it as an `ExtResource` in a `tscn` or a `tres` file. Loading the long way will make sure that you get the cached version and thus that it gets saved as an `ExtResource` in the scene or resource file. Maybe I'll write a helper for newing up resources from c#...)

**Comment by cgbeutler** [comment]
To create a C# custom resource saved to `FileSystem` simply:

Right-click > New Resource...
Choose "Resource" (The base type. Should be first in the list.)
Hit "Create"
Name it as usual. Click "Save".
Double-click the new resource in the `FileSystem` dock to bring up its inspector.
Find the `Script` field and drag-drop or `load` your `.cs` script.
![image](https://user-images.githubusercontent.com/9751923/118748396-f5f2fb80-b818-11eb-9ba4-75fa4feb466a.png)
Use Ctrl-s to save those changes. You should now see all the exported fields.
![image](https://user-images.githubusercontent.com/9751923/118748610-3eaab480-b819-11eb-855d-047a85c74109.png)

Note that if you change a script, godot will create a new object for the "new" script, then copy all exported variables to matching ones in the updated script. Any info that doesn't line up, will be lost.

**Comment by DJ-Dav** [comment]
Could a workaround be to add a child node with the properties of the c# resource and use setget to update the node when the resource's properties are changed, and add a GD script to the node to update the resource when it's properties are edited in the inspector? I am yet to try this out, but could this be a possible solution.

**Comment by cgbeutler** [comment]
@DJ-Dav, yeah, but that would involve a lot of boilerplate and remove a lot of the benefits of resources. Godot basically has 2 garbage management methods: The tree and resources. The tree will clean up stuff when that branch is deleted. When you need data in multiple places in the tree, like a sprite or something, resources allow that cross-tree data, caching it for you and cleaning it up when all references to it are gone. If you only need some data in only one place in the tree, you don't really need a custom resource, saving a node to a scene would do just fine in those cases.
In other words, nodes and resources are designed to solve different data dependency needs. You could work up a mix, but that kinda kills the benefits of the design of Godot.

**Comment by ducabyte** [comment]
Still not implemented in 3.4.3 stable Mono official

**Comment by willnationsdev** [comment]
@ducabyte you just have to keep an eye on #48201 and its 3.x counterpart #44879 since those are the PRs that add support for this. Not sure if they specifically will be merged in, but either they or a separate PR that is derived from their contents will likely be merged in before the 4.0 release (core team has expressed an intent to add the feature for 4.0), and then it *may* end up being backported to a subsequent 3.x release after that.

**Comment by ducabyte** [comment]
> @ducabyte you just have to keep an eye on #48201 and its 3.x counterpart #44879 since those are the PRs that add support for this. Not sure if they specifically will be merged in, but either they or a separate PR that is derived from their contents will likely be merged in before the 4.0 release (core team has expressed an intent to add the feature for 4.0), and then it _may_ end up being backported to a subsequent 3.x release after that.

ty @willnationsdev

**Comment by awardell** [comment]
@cgbeutler I tried your solution, but I'm not seeing script variables show up in the inspector. Adding [Tool] didn't resolve things. Any idea what I could try to get it working? I'm on 3.4.4.stable.mono.

![image](https://user-images.githubusercontent.com/1361278/183745676-64aa1fcd-751b-4286-948a-a58127b940c6.png)

![image](https://user-images.githubusercontent.com/1361278/183745814-96283535-b185-4b9c-8e92-a7ab5d47ad39.png)

**Comment by cgbeutler** [comment]
@awardell I think it has to be a property, not a field...?
```
using Godot;

[Tool]
public class HexCellData : Resource
{
  [Export]
  private string testStr { get; set; } = "test";
}
```
Also, if there are any compilation errors _anywhere_, the new c# exports won't show up till it compiles.

**Comment by awardell** [comment]
I just discovered it was my own stupid operator error. I needed to build solution first. Now it works how I have it, as a field, even without [Tool]. My apologies for the distraction! Thanks for your reply.

**Comment by kaluluosi** [comment]
This problem still there?

**Comment by DJ-Dav** [comment]
> This problem still there?

I don't know,  i am having issues getting .NET stuff to work these days. I will test this on a different machine later.

**Comment by mazingtech** [comment]
I'm struggling with custom resource in C# too, they should really add a note: C# support is not good enough to make a game now, just for fun only! And the performance on MacOS is totally a mess, please wait until we come up a new version in the future!

Totally a waste of time.

**Comment by raulsntos** [comment]
- Fixed by https://github.com/godotengine/godot/pull/72619.
- Use the `[GlobalClass]` attribute, see the documentation about [_C# global classes_](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_global_classes.html).

**Comment by nubbyninja** [comment]
Addendum: If custom types still aren't showing up after adding [GlobalClass], click Build, then try again.

**Comment by felipemullen** [comment]
> I'm struggling with custom resource in C# too, they should really add a note: C# support is not good enough to make a game now, just for fun only! And the performance on MacOS is totally a mess, please wait until we come up a new version in the future!
> 
> Totally a waste of time.

I am on MacOS M1 chip using dotnet 7.0 and not seeing any issues. Godot version 4.1.1
